external/hive-delta/src/main/java/io/delta/hive/DeltaStorageHandler.java

```diff
@@ -0,0 +1,203 @@
+                result.put(condition.getColumnDesc().getColumn(), conditions);
+            }
+            conditions.add(condition);
+        }
+        return result;
+    }
+
+    private static ExprNodeGenericFuncDesc extractResidualCondition(
+            IndexPredicateAnalyzer analyzer,
+            List<IndexSearchCondition> searchConditions, ExprNodeGenericFuncDesc inputExpr) {
+        if (inputExpr == null) {
+            return analyzer.translateOriginalConditions(searchConditions);
+        }
+        List<ExprNodeDesc> children = new ArrayList<ExprNodeDesc>();
+        children.add(analyzer.translateOriginalConditions(searchConditions));
+        children.add(inputExpr);
+        return new ExprNodeGenericFuncDesc(TypeInfoFactory.booleanTypeInfo,
+                FunctionRegistry.getGenericUDFForAnd(), children);
+    }
+
+    private static ExprNodeGenericFuncDesc extractStorageHandlerCondition(
+            IndexPredicateAnalyzer analyzer,
+            List<IndexSearchCondition> searchConditions, ExprNodeGenericFuncDesc inputExpr) {
+        if (inputExpr == null) {
+            return analyzer.translateSearchConditions(searchConditions);
+        }
+        List<ExprNodeDesc> children = new ArrayList<ExprNodeDesc>();
+        children.add(analyzer.translateSearchConditions(searchConditions));
+        children.add(inputExpr);
+        return new ExprNodeGenericFuncDesc(TypeInfoFactory.booleanTypeInfo,
+                FunctionRegistry.getGenericUDFForAnd(), children);
+    }
+
+    public HiveMetaHook getMetaHook() {
+        return this;
+    }
+
+    public void preCreateTable(Table tbl) throws MetaException {
+        boolean isExternal = MetaStoreUtils.isExternalTable(tbl);
+        if (!isExternal) {
+            throw new MetaException("HiveOnDelta should be an external table.");
+        } else if (tbl.getPartitionKeysSize() > 0) {
+            throw new MetaException("HiveOnDelta does not support to create a partition hive table");
+        }
+
+        Map<String, String> serdeParam = tbl.getSd().getSerdeInfo().getParameters();
+        String deltaRootString = serdeParam.get(DELTA_TABLE_PATH);
+
+        try {
+            if (deltaRootString == null || deltaRootString.trim().length() == 0) {
+                throw new MetaException("Error: delta.table.path is missing in SERDEPROPERTIES when creating table");
+            } else {
+                Path deltaPath = new Path(deltaRootString);
+                FileSystem fs = deltaPath.getFileSystem(getConf());
+                if (!fs.exists(deltaPath)) {
+                    throw new MetaException("delta.table.path(" + deltaRootString + ") does not exist...");
+                } else {
+                    Map<String, String> partitionProps = JavaConversions.mapAsJavaMap(
+                            DeltaHelper.checkHiveColsInDelta(deltaPath, tbl.getSd().getCols()));
+                    tbl.getSd().getSerdeInfo().getParameters().putAll(partitionProps);
```

Comment:
@zsxwing can Delta table change partition columns?

Reply:
if it can then we need to infer the partition columns everytime we read the delta hive serde table.